firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift

```diff
@@ -819,56 +824,101 @@ class LegacyTabManager: NSObject, FeatureFlaggable, TabManager, TabEventHandler
+    private func updateSelectedTabAfterRemovalOf(_ removedTab: Tab, deletedIndex: Int) {
+        // If the currently selected tab has been deleted, try to select the next most reasonable tab.
+        if deletedIndex == _selectedIndex {
+            // First, check if the user has closed the last viable tab of the current browsing mode: private or normal.
+            // If so, handle this gracefully (i.e. close the last private tab should open the most recent normal active tab).
+            let viableTabs = removedTab.isPrivate
+                                 ? privateTabs
+                                 : normalActiveTabs // We never want to surface an inactive tab
+            guard !viableTabs.isEmpty else {
+                // If the selected tab is closed, and is private browsing, try to select a recent normal active tab. For all
+                // other cases, open a new normal active tab.
+                if removedTab.isPrivate,
+                   let mostRecentActiveTab = mostRecentTab(inTabs: normalActiveTabs) {
+                    selectTab(mostRecentActiveTab, previous: removedTab)
                 } else {
-                    let mostRecentTab = mostRecentTab(inTabs: activeViableTabs) ?? activeViableTabs.last
-                    selectTab(mostRecentTab, previous: tab)
+                    selectTab(addTab(), previous: removedTab)
                 }
+                return
+            }
+
+            if let mostRecentViableTab = mostRecentTab(inTabs: viableTabs), mostRecentViableTab == removedTab.parent {
+                // 1. Try to select the most recently used viable tab, if it's the removed tab's parent.
+                selectTab(mostRecentViableTab, previous: removedTab)
+            } else if !removedTab.isNormalAndInactive,
+                      let rightOrLeftTab = findRightOrLeftTab(forRemovedTab: removedTab, withDeletedIndex: deletedIndex) {
+                // 2. Try to select an array neighbour of the same tab type, except if the removed tab is inactive (unlikely
+                // edge case).
+                selectTab(rightOrLeftTab, previous: removedTab)
+            } else {
+                // 3. If there are no suitable active tabs to select, create a new normal active tab.
+                // (Note: It's possible to fall into here when all tabs have become inactive, especially when debugging.)
+                selectTab(addTab(), previous: removedTab)
             }
         } else if deletedIndex < _selectedIndex {
-            // If we delete an active tab that's before the selected active tab, we need to shift our selection index
-            // since the array size has changed.
-            let selected = tabs[safe: _selectedIndex - 1]
-            selectTab(selected, previous: selected)
+            // If we delete a tab in the `tabs` array that's earlier than the selected tab, we need to shift our index.
```

Comment:
"earlier than" -> "ahead of" maybe?  Unless I'm misunderstanding the comment

Reply:
Sure, that works! 👍 Committed.